packages/headless/src/controllers/commerce/facets/core/headless-core-facet-generator.ts

```diff
@@ -0,0 +1,73 @@
+import {CommerceEngine} from '../../../../app/commerce-engine/commerce-engine';
+import {facetOrderReducer as facetOrder} from '../../../../features/facets/facet-order/facet-order-slice';
+import {
+  commerceFacetSetReducer as commerceFacetSet
+} from '../../../../features/commerce/facets/facet-set/facet-set-slice';
+import {
+  CommerceFacetSetSection,
+  FacetOrderSection,
+} from '../../../../state/state-sections';
+import {loadReducerError} from '../../../../utils/errors';
+import {
+  buildController,
+  Controller,
+} from '../../../controller/headless-controller';
+import {
+  Facet,
+  FacetBuilder,
+} from './headless-core-facet';
+
+export interface FacetGenerator extends Controller {
+  state: FacetGeneratorState;
+}
+
+export interface FacetGeneratorState {
+  facets: Facet[];
+}
+
+export interface FacetGeneratorOptions {
+  buildFacet: FacetBuilder;
+}
+
+export function buildCoreFacetGenerator(engine: CommerceEngine, options: FacetGeneratorOptions): FacetGenerator {
+  if (!loadFacetGeneratorReducers(engine)) {
+    throw loadReducerError;
+  }
+
+  const controller = buildController(engine);
+
+  const createFacet = (field: string) => {
+    const {type} = engine.state.commerceFacetSet[field].request
+
+    switch (type) {
+      case 'regular':
+      default:
+        // eslint-disable-next-line @cspell/spellchecker
+        // TODO CAPI-90, CAPI-91: Use different controllers for different facet types taken from facet.type
+        return options.buildFacet(engine, {
+          options: {
+            field,
+          }
+        });
+    }
+  };
+
+  return {
+    ...controller,
+
+    get state() {
+      return {
+        facets: engine.state.facetOrder.map(createFacet) ?? [],
```

Comment:
A thought:

As a data structure, the JavaScript `Map` object remembers the order of insertion (see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map).

Could we define the `CommercefacetSetState` type as a `Map<string, CommerceFacetSlice>` instead of a `Record<string, CommerceFacetSlice>` to avoid having to duplicate facets in a `facetOrder` array?